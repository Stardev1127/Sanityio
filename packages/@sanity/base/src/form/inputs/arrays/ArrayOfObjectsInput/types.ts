import type {ComponentType} from 'react'
import {KeyedSegment, Path} from '@sanity/types'

export type ModalType = 'modal' | 'fullscreen' | string

export type ArrayMember = {
  _type?: string
  _key: string
  [key: string]: any
}

/**
 *  @deprecated
 *  */
export interface InsertEvent {
  item: ArrayMember
  position: 'before' | 'after'
  referenceItem: KeyedSegment | number
  edit?: boolean
}
export type ReferenceItemComponentType = ComponentType<any>
