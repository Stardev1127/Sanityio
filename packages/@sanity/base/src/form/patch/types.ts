/* eslint-disable camelcase */

import {Path} from '@sanity/types'

/**
 * @internal
 */
export type FormPatchJSONValue =
  | number
  | string
  | boolean
  | {[key: string]: FormPatchJSONValue}
  | FormPatchJSONValue[]

/**
 * @internal
 */
export type FormPatchOrigin = 'remote' | 'local' | 'internal'

/**
 * @internal
 */
export type FormSetPatch = {
  path: Path
  type: 'set'
  origin?: FormPatchOrigin
  value: FormPatchJSONValue
}

/**
 * @internal
 */
export type FormIncPatch = {
  path: Path
  type: 'inc'
  origin?: FormPatchOrigin
  value: FormPatchJSONValue
}

/**
 * @internal
 */
export type FormDecPatch = {
  path: Path
  type: 'dec'
  origin?: FormPatchOrigin
  value: FormPatchJSONValue
}

/**
 * @internal
 */
export type FormSetIfMissingPatch = {
  path: Path
  origin?: FormPatchOrigin
  type: 'setIfMissing'
  value: FormPatchJSONValue
}

/**
 * @internal
 */
export type FormUnsetPatch = {
  path: Path
  origin?: FormPatchOrigin
  type: 'unset'
}

/**
 * @internal
 */
export type FormInsertPatchPosition = 'before' | 'after'

/**
 * @internal
 */
export type FormInsertPatch = {
  path: Path
  origin?: FormPatchOrigin
  type: 'insert'
  position: FormInsertPatchPosition
  items: FormPatchJSONValue[]
}

/**
 * @internal
 */
export type FormDiffMatchPatch = {
  path: Path
  type: 'diffMatchPatch'
  origin?: FormPatchOrigin
  value: string
}

/**
 * @internal
 */
export type FormPatch =
  | FormSetPatch
  | FormSetIfMissingPatch
  | FormUnsetPatch
  | FormInsertPatch
  | FormDiffMatchPatch
