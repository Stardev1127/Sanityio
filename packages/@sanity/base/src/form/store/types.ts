import {
  ArraySchemaType,
  BooleanSchemaType,
  NumberSchemaType,
  ObjectSchemaType,
  Path,
  StringSchemaType,
  ValidationMarker,
} from '@sanity/types'
import * as React from 'react'
import {ComponentType} from 'react'
import {PatchEvent} from '../patch'
import {FormFieldPresence} from '../../presence'
import {ObjectInputProps} from './formState'

export interface StateTree<T> {
  value: T | undefined
  children?: {
    [key: string]: StateTree<T>
  }
}

export interface FieldGroup {
  name: string
  title?: string
  icon?: ComponentType<void>
  default?: boolean
  selected?: boolean
  disabled?: boolean
}

export type ObjectMember = FieldMember | FieldSetMember

// note: array members doesn't have the field/fieldSet divide
export type ArrayMember = ObjectInputProps

export interface FieldMember {
  type: 'field'
  field: FieldProps
  key: string
}

export interface FieldSetProps {
  name: string
  title?: string
  hidden?: boolean
  collapsible?: boolean
  collapsed?: boolean
  onSetCollapsed: (collapsed: boolean) => void
  fields: FieldMember[]
}

export interface FieldSetMember {
  type: 'fieldSet'
  key: string
  fieldSet: FieldSetProps
}

interface BaseFieldProps {
  name: string
  title?: string
  description?: string
  // id is a stringified version of the path
  id: string
  // the full content path of this input
  path: Path
  focused: boolean
  presence: FormFieldPresence[]
  validation: ValidationMarker[]
  index: number
  level: number
  hidden?: boolean
  readOnly?: boolean
  onChange: (patchEvent: PatchEvent) => void
  onFocus: (focusEvent: React.FocusEvent) => void
}

export interface StringFieldProps extends BaseFieldProps {
  kind: 'string'
  type: StringSchemaType
  value?: string
}

export interface NumberFieldProps extends BaseFieldProps {
  kind: 'number'
  type: NumberSchemaType
  value?: number
}

export interface BooleanFieldProps extends BaseFieldProps {
  kind: 'boolean'
  type: BooleanSchemaType
  value?: boolean
}

export interface ObjectFieldProps extends BaseFieldProps {
  kind: 'object'
  type: ObjectSchemaType
  members: ObjectMember[]
  groups?: FieldGroup[]
  onSelectGroup: (name: string) => void
  onFocus: (focusEvent: React.FocusEvent) => void
  onBlur: (focusEvent: React.FocusEvent) => void
  focusPath: Path
  hidden?: boolean
  value?: Record<string, unknown>
  readOnly?: boolean
  collapsed?: boolean
  collapsible?: boolean
  onSetCollapsed: (collapsed: boolean) => void
}

export interface ArrayFieldProps extends BaseFieldProps {
  kind: 'array'
  type: ArraySchemaType
  members: ObjectInputProps[]
  focusPath: Path
  value?: unknown[]
}

export type FieldProps =
  | StringFieldProps
  | ObjectFieldProps
  | ArrayFieldProps
  | NumberFieldProps
  | BooleanFieldProps
