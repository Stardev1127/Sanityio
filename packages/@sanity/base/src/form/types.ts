import {
  ArraySchemaType,
  BooleanSchemaType,
  NumberSchemaType,
  ObjectField,
  ObjectSchemaType,
  ObjectSchemaTypeWithOptions,
  Path,
  SchemaType,
  StringSchemaType,
  ValidationMarker,
} from '@sanity/types'
import {FormFieldPresence} from '../presence'
import {PatchEvent} from './patch'

export type FIXME = any

/**
 * Export `PortableTextMarker` so it can be used to build custom Portable Text markers.
 *
 * @alpha
 */
export type RenderCustomMarkers = (markers: PortableTextMarker[]) => React.ReactNode

/**
 * A generic marker for attaching metadata to specific nodes of the Portable Text input.
 *
 * @alpha
 */
export interface PortableTextMarker {
  type: string
  data?: unknown
  path: Path
}

/**
 * @alpha
 */
export type FormBuilderArrayFunctionComponent = React.ComponentType<
  FormArrayInputFunctionsProps<any, any>
>

/**
 * @alpha
 */
export type FormBuilderCustomMarkersComponent = React.ComponentType<{markers: PortableTextMarker[]}>

/**
 * @alpha
 */
export type FormBuilderMarkersComponent = React.ComponentType<{
  markers: PortableTextMarker[]
  renderCustomMarkers?: RenderCustomMarkers
  validation: ValidationMarker[]
}>

/**
 * @alpha
 */
export type FormBuilderInputComponentMap = Record<
  string,
  React.ComponentType<FormInputProps<any, any>> | undefined
>

/**
 * These are the props an implementation of the ArrayFunctions component will receive
 */
export interface FormArrayInputFunctionsProps<SchemaType extends ArraySchemaType, MemberType> {
  children?: React.ReactNode
  className?: string
  onAppendItem: (itemValue: MemberType) => void
  onChange: (event: PatchEvent) => void
  onCreateValue: (type: SchemaType) => MemberType
  onFocusItem: (item: MemberType, index: number) => void
  onPrependItem: (itemValue: MemberType) => void
  readOnly?: boolean
  type: SchemaType
  value?: MemberType[]
}

/**
 * @alpha
 */
export interface FormBuilderFilterFieldFn {
  (type: ObjectSchemaTypeWithOptions, field: ObjectField, selectedLanguageIds: string[]): boolean
}

/**
 * @alpha
 */
export type FormInputProps<
  T = any,
  S = T extends Array<any>
    ? ArraySchemaType
    : T extends boolean
    ? BooleanSchemaType
    : T extends string
    ? StringSchemaType
    : T extends number
    ? NumberSchemaType
    : T extends Record<string, any>
    ? ObjectSchemaType
    : SchemaType
> = {
  compareValue?: T
  focusPath?: Path
  level: number
  validation: ValidationMarker[]
  onBlur?: () => void
  // @todo allow implementers to pass a native DOM ChangeEvent
  // @todo allow implementers to pass an array of patch objects
  // @todo allow implementers to simply pass the new value
  // @todo deprecate `PatchEvent`
  onChange: (event: PatchEvent) => void
  // NOTE: we should allow implementers of custom inputs to forward the passed onFocus to native
  // element's onFocus handler, but use Path consistently on internal inputs
  onFocus: (pathOrEvent?: Path | React.FocusEvent<any>) => void
  presence: FormFieldPresence[]
  // @todo: should the `readOnly` use the `ConditionalProperty` type?
  // readOnly?: ConditionalProperty
  readOnly?: boolean
  type: S
  value?: T
}

/**
 * @alpha
 */
export type FormInputComponentResolver = (
  type: SchemaType
) => React.ComponentType<FormInputProps<any, any>> | null | false | undefined

/**
 * @alpha
 */
export interface FormPreviewProps {
  actions?: React.ReactNode
  fallbackTitle?: React.ReactNode
  layout?: string
  type: SchemaType
  value?: Array<unknown> | Record<string, unknown> | number | boolean | string
  withBorder?: boolean
  withRadius?: boolean
}

/**
 * @alpha
 */
export type FormPreviewComponentResolver = (
  type: SchemaType
) => React.ComponentType<FormPreviewProps> | null | false | undefined
