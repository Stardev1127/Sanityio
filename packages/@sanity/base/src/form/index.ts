// This exports the public api of '@sanity/form-builder'

// sanity
export * from './sanity/contexts'
export {SanityFormBuilder} from './sanity/SanityFormBuilder'
export {SanityFormBuilderProvider} from './sanity/SanityFormBuilderProvider'
export {checkoutPair} from './sanity/formBuilderValueStore'
export {HashFocusManager} from './sanity/focusManagers/HashFocusManager'
export {SimpleFocusManager} from './sanity/focusManagers/SimpleFocusManager'

// utils
export {withDocument} from './utils/withDocument'
export {withValuePath} from './utils/withValuePath'

// root components
export type {FormBuilderInputProps} from './FormBuilderInput'
export {FormBuilderInput, FormBuilderInputInstance} from './FormBuilderInput'

// inputs
export {PortableTextInput as BlockEditor} from './inputs/PortableText/PortableTextInput'

export * from './patchChannel'
export * from './sanity'

export * from './patch/PatchEvent'
export * from './patch/patch'
export * from './patch/types'
export * from './types'
export type {MutationPatch} from './utils/mutationPatch'
export {fromMutationPatches, toMutationPatches} from './utils/mutationPatch'
export {decodePath, encodePath} from './utils/path'
