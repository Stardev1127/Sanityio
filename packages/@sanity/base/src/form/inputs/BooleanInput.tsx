/* eslint-disable import/no-unresolved */

import React, {useCallback} from 'react'
import styled from 'styled-components'
import {useId} from '@reach/auto-id'
import {Box, Card, Checkbox, Flex, Switch} from '@sanity/ui'
import {BooleanSchemaType} from '@sanity/types'
import {FormFieldHeaderText, FormFieldStatus} from '../../components/formField'
import {FieldPresence} from '../../presence'
import {FormInputProps} from '../types'
import {PatchEvent, set} from '../patch'
import {ChangeIndicator} from '../../components/changeIndicators'

export type BooleanInputProps = FormInputProps<boolean, BooleanSchemaType>

const CenterAlignedBox = styled(Box)`
  align-self: center;
`

const ZeroLineHeightBox = styled(Box)`
  line-height: 0;
`

export const BooleanInput = React.forwardRef(function BooleanInput(
  props: BooleanInputProps,
  ref: React.ForwardedRef<HTMLInputElement>
) {
  const {onChange} = props
  const {value, type, readOnly, onFocus, validation, presence} = props
  const layout = type.options?.layout || 'switch'
  const inputId = useId()

  const handleChange = useCallback(
    (event: React.SyntheticEvent<HTMLInputElement>) => {
      onChange(PatchEvent.from(set(event.currentTarget.checked)))
    },
    [onChange]
  )

  const indeterminate = typeof value !== 'boolean'
  const checked = value || false

  const LayoutSpecificInput = layout === 'checkbox' ? Checkbox : Switch

  return (
    <ChangeIndicator>
      <Card border radius={1}>
        <Flex>
          <ZeroLineHeightBox padding={3}>
            <LayoutSpecificInput
              id={inputId}
              ref={ref}
              label={type.title}
              readOnly={Boolean(readOnly)}
              onChange={handleChange}
              onFocus={onFocus}
              indeterminate={indeterminate}
              checked={checked}
              style={{margin: -4}}
              disabled={readOnly}
            />
          </ZeroLineHeightBox>
          <Box flex={1} paddingY={3}>
            <FormFieldHeaderText
              description={type.description}
              inputId={inputId}
              validation={validation}
              title={type.title}
            />
          </Box>
          <CenterAlignedBox paddingX={3} paddingY={1}>
            <FormFieldStatus maxAvatars={1} position="top">
              <FieldPresence maxAvatars={1} presence={presence} />
            </FormFieldStatus>
          </CenterAlignedBox>
        </Flex>
      </Card>
    </ChangeIndicator>
  )
})
