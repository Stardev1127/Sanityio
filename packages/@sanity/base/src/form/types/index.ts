export * from './event'
export * from './inputProps'
export * from './fieldProps'
export * from './itemProps'
export * from './renderCallback'
export * from './_transitional'
export * from './fieldGroups'
export * from './fieldProps'
export * from '../store/types/members'
export * from '../store/types/state'
