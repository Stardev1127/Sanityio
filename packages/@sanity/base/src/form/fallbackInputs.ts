import {ObjectInput} from './inputs/ObjectInput'
import {ArrayInput} from './inputs/arrays/ArrayOfObjectsInput'
import {BooleanInput} from './inputs/BooleanInput'
import {NumberInput} from './inputs/NumberInput'
import {TextInput} from './inputs/TextInput'

export const fallbackInputs = {
  object: {input: ObjectInput},
  array: {input: ArrayInput},
  boolean: {input: BooleanInput},
  number: {input: NumberInput},
  string: {input: TextInput},
}
