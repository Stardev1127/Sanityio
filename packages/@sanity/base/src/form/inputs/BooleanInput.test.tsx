import userEvent from '@testing-library/user-event'
import {defineType} from '@sanity/types'
import React from 'react'
import {renderBooleanInput} from '../../../test/form/renderBooleanInput'
import {BooleanInput} from './BooleanInput'

const defs = {
  booleanTest: defineType({
    name: 'booleanTest',
    title: 'Switch',
    type: 'boolean',
  }),

  booleanReadOnly: defineType({
    name: 'booleanReadOnly',
    title: 'Read-only',
    type: 'boolean',
    readOnly: true,
  }),

  readOnlyCallback: defineType({
    name: 'readOnlyCallback',
    title: 'Boolean with callback',
    type: 'boolean',
    readOnly: () => false,
  }),

  readOnlyWithDocument: defineType({
    name: 'readOnlyWithDocument',
    title: 'Boolean read-only with document',
    type: 'boolean',
    readOnly: (context) => context.document?.title === 'Hello world',
  }),

  booleanHidden: {
    name: 'booleanHidden',
    title: 'Hidden',
    type: 'boolean',
    hidden: true,
  },

  hiddenCallback: {
    name: 'hiddenCallback',
    title: 'Boolean with callback',
    type: 'boolean',
    hidden: () => false,
  },

  hiddenWithDocument: {
    name: 'hiddenWithDocument',
    title: 'Boolean hidden with document',
    type: 'boolean',
    hidden: ({document}: any) => document.title === 'Hello world',
  },
}

it('renders the boolean input field', () => {
  const {result} = renderBooleanInput({
    fieldDefinition: defs.booleanTest,
    render: (inputProps) => <BooleanInput {...inputProps} />,
  })

  const input = result.container.querySelector('input[id="booleanTest"]')
  expect(input).toBeDefined()
  expect(input).toHaveAttribute('type', 'checkbox')
  expect(input).toBePartiallyChecked()
})

describe('Mouse accessibility', () => {
  it('emits onFocus when clicked', () => {
    const {onFocus, result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })
    const input = result.container.querySelector('input[id="booleanTest"]')
    userEvent.click(input!)
    expect(onFocus).toBeCalled()
    expect(onFocus.mock.calls).toMatchSnapshot()
  })

  it('emits onChange when clicked', () => {
    const {onChange, result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="booleanTest"]')
    userEvent.click(input!)
    expect(onChange).toBeCalled()
    expect(onChange.mock.calls).toMatchSnapshot()
  })
})

describe('Keyboard accessibility', () => {
  it('emits onFocus when tabbing to input', () => {
    const {onFocus, result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="booleanTest"]')
    userEvent.tab({focusTrap: result.container})
    expect(input).toHaveFocus()
    expect(onFocus).toBeCalled()
    expect(onFocus.mock.calls).toMatchSnapshot()
  })

  it('emits onChange when pressing enter', () => {
    const {onChange, result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    userEvent.tab({focusTrap: result.container})
    userEvent.keyboard('{space}')
    expect(onChange).toBeCalled()
    expect(onChange.mock.calls).toMatchSnapshot()
  })

  it('emits onBlur when navigating away from field', () => {
    const {onBlur, result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="booleanTest"]')
    userEvent.tab({focusTrap: result.container})
    userEvent.tab()
    expect(input).not.toHaveFocus()

    expect(onBlur).toBeCalled()
    expect(onBlur.mock.calls).toMatchSnapshot()
  })
})

describe('Layout options', () => {
  it('renders a switch (default)', () => {
    const {result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="booleanTest"][data-ui="Switch"]')
    expect(input).toBeDefined()
  })

  it('renders a checkbox', () => {
    const {result} = renderBooleanInput({
      fieldDefinition: defs.booleanTest,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="booleanTest"][data-ui="Checkbox"]')
    expect(input).toBeDefined()
  })
})

describe('readOnly property', () => {
  it('makes field read-only', () => {
    const {onChange, result} = renderBooleanInput({
      fieldDefinition: defs.booleanReadOnly,
      render: (inputProps) => <BooleanInput {...inputProps} readOnly />,
    })

    const input = result.container.querySelector('input[id="booleanReadOnly"]')
    expect(input).toBeDisabled()

    // Mouse event
    userEvent.click(input!)
    // expect(input).toHaveFocus()
    expect(onChange).not.toBeCalled()

    // Keyboard event
    userEvent.tab({focusTrap: result.container})
    expect(input).not.toHaveFocus()
  })

  it('does not make field read-only with callback', () => {
    const {onChange, result} = renderBooleanInput({
      fieldDefinition: defs.readOnlyCallback,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="readOnlyCallback"]')
    expect(input).not.toBeDisabled()

    // Mouse event
    userEvent.click(input!)
    expect(onChange).toBeCalled()

    // Keyboard event
    userEvent.tab({shift: true})
    userEvent.tab({focusTrap: result.container})
    userEvent.keyboard('{space}')
    expect(onChange).toBeCalled()
    expect(onChange.mock.calls).toMatchSnapshot()
  })

  it.skip('makes field read-only based on value in document', () => {
    const {onChange, result} = renderBooleanInput({
      fieldDefinition: defs.readOnlyWithDocument,
      props: {documentValue: {title: 'Hello, world'}},
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })

    const input = result.container.querySelector('input[id="readOnlyWithDocument"]')
    expect(input).toBeDisabled()

    // Mouse event
    userEvent.click(input!)
    expect(onChange).not.toBeCalled()

    // Keyboard event
    userEvent.tab({focusTrap: result.container})
    expect(input).not.toHaveFocus()
  })
})

describe('hidden property', () => {
  it('hides field', () => {
    expect(() =>
      renderBooleanInput({
        fieldDefinition: defs.booleanHidden,
        render: (inputProps) => <BooleanInput {...inputProps} />,
      })
    ).toThrow('no field member: booleanHidden')
  })

  it('does not hide field with callback', () => {
    const {result} = renderBooleanInput({
      fieldDefinition: defs.hiddenCallback,
      render: (inputProps) => <BooleanInput {...inputProps} />,
    })
    const input = result.container.querySelector('input[id="hiddenCallback"]')

    expect(input).toBeDefined()
  })

  it.skip('hides field based on value in document', () => {
    expect(() =>
      renderBooleanInput({
        fieldDefinition: defs.hiddenWithDocument,
        props: {documentValue: {title: 'Hello, world'}},
        render: (inputProps) => <BooleanInput {...inputProps} />,
      })
    ).toThrow('no field member: hiddenWithDocument')
  })
})
