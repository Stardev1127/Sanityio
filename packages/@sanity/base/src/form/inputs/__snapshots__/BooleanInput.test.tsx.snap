// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard accessibility emits onChange when pressing enter 1`] = `
Array [
  Array [
    Array [
      Object {
        "set": Object {
          "booleanTest": true,
        },
      },
    ],
  ],
]
`;

exports[`Keyboard accessibility emits onFocus when navigating to field 1`] = `
Array [
  Array [
    Array [
      "booleanTest",
    ],
  ],
]
`;

exports[`Mouse accessibility emits onChange when clicked 1`] = `
Array [
  Array [
    Array [
      Object {
        "set": Object {
          "booleanTest": true,
        },
      },
    ],
  ],
]
`;

exports[`Mouse accessibility emits onFocus when clicked 1`] = `
Array [
  Array [
    Array [
      "booleanTest",
    ],
  ],
]
`;

exports[`readOnly property does not make field read-only with callback 1`] = `
Array [
  Array [
    Array [
      Object {
        "set": Object {
          "readOnlyCallback": true,
        },
      },
    ],
  ],
  Array [
    Array [
      Object {
        "set": Object {
          "readOnlyCallback": true,
        },
      },
    ],
  ],
]
`;
